drivers/can/can_esp32_twai.c

```diff
@@ -90,6 +179,18 @@ static int can_esp32_twai_init(const struct device *dev)
 		return err;
 	}
 
+#ifndef CONFIG_SOC_ESP32
+	/*
+	 * TWAI_CLOCK_DIVIDER_REG is incompatible with CAN_SJA1000_CDR for non-ESP32 MCUs
+	 *   - TWAI_CD has length of 8 bits instead of 3 bits
+	 *   - TWAI_CLOCK_OFF at BIT(8) instead of BIT(3)
+	 *   - TWAI_EXT_MODE bit missing (always "extended" = PeliCAN mode)
+	 *
+	 * Resetting the register to 0 to remove the the mode bit set in can_sja_1000_init().
+	 */
+	can_esp32_twai_write_reg32(dev, CAN_SJA1000_CDR, 0U);
```

Comment:
Hardcoding this to `0U` removes the possibility of turning off the clock output or setting the divisor?

Reply:
As far as I understand this register is useless for ESP32, so I reset the clock divider to zero (as done in the SJA1000 backend as well).

The *external* CLKOUT configured with this register could be used for an external MCU in case of an actual SJA1000 chip. But it doesn't really seem to make sense if the SJA1000 is built directly *into* the SOC.

The CLKOUT is also more or less undocumented in the ESP32-C3 reference manual. If you search for it, the only hit (except for the register definition itself) is a table which says that the signal is not muxed to any pin.

Reply:
The current TWAI driver sets the CDR register based on the `clkout-divider` DTS property obtained via the `CAN_ESP32_TWAI_DT_CDR_INST_GET()` macro. (see https://github.com/zephyrproject-rtos/zephyr/blob/98580f1c22481c200db2f37881e9c5161f2bd35f/dts/bindings/can/espressif%2Cesp32-twai.yaml#L26 and https://github.com/zephyrproject-rtos/zephyr/blob/98580f1c22481c200db2f37881e9c5161f2bd35f/drivers/can/can_esp32_twai.c#L119).

Pinctrl for muxing the CLKOUT signal for ESP32* was added in https://github.com/zephyrproject-rtos/hal_espressif/pull/108 and fully supported.

While using this signal may or may not make sense for your use-case, I would like to retain support for this no matter if the driver is used on the original ESP32 or a later variant.

Reply:
Interesting. I still can't find this feature being documented anywhere in the ESP32 and ESP32-C3 reference manual.

Digging deeper into the Espressif HAL reveals that the divider can now be 1 or an even number from 2 to 490 (!):

https://github.com/zephyrproject-rtos/hal_espressif/blob/68e62d9989e8a8b250348ef28aca1c70da00b0c1/components/hal/esp32c3/include/hal/twai_ll.h#L643-L665

I think your previous approach with a dedicated `#define` to translate divider to register doesn't make sense anymore, as I don't want to create 490/2 defines. Also the enum in the binding would get very long...
Any suggestion how to fix this in a clean way? Remove the enum restricting the values in the binding and do a check + calculation similar to the Espressif HAL in `can_esp32_twai_init`?

Reply:
Agreed. Using `BUILD_ASSERT` sounds like a good solution.

Reply:
Remember to update the binding documentation as well.

Reply:
The macros are getting a bit convoluted, but it's finally still possible to evaluate the `clkout-divider` at compile-time. Please have a look again. Binding doc is also updated.